edb/server/compiler/compiler.py

```diff
@@ -1410,38 +1472,211 @@ def _compile_ql_migration(
+                    tx_action=tx_action,
+                    cacheable=False,
+                    modaliases=None,
+                    single_unit=True,
+                    user_schema=ctx.state.current_tx().get_user_schema(),
+                    cached_reflection=(
+                        current_tx.get_cached_reflection_if_updated()
+                    )
+                )
+
+        elif isinstance(ql, qlast.AbortMigration):
+            mstate = self._assert_in_migration_block(ctx, ql)
 
             if mstate.initial_savepoint:
-                current_tx.commit_migration(mstate.initial_savepoint)
-                sql = ddl_query.sql
+                current_tx.abort_migration(mstate.initial_savepoint)
+                sql = (b'SELECT LIMIT 0',)
                 tx_action = None
             else:
-                tx_cmd = qlast.CommitTransaction()
+                tx_cmd = qlast.RollbackTransaction()
                 tx_query = self._compile_ql_transaction(ctx, tx_cmd)
-                sql = ddl_query.sql + tx_query.sql
+                sql = tx_query.sql
                 tx_action = tx_query.action
 
+            current_tx.update_migration_state(None)
             query = dbstate.MigrationControlQuery(
                 sql=sql,
-                ddl_stmt_id=ddl_query.ddl_stmt_id,
+                action=dbstate.MigrationAction.ABORT,
+                tx_action=tx_action,
+                cacheable=False,
+                modaliases=None,
+                single_unit=True,
+            )
+
+        elif isinstance(ql, qlast.DropMigration):
+            self._assert_not_in_migration_block(ctx, ql)
+
+            query = self._compile_and_apply_ddl_stmt(ctx, ql)
+
+        elif isinstance(ql, qlast.StartMigrationRewrite):
+            self._assert_not_in_migration_block(ctx, ql)
+            self._assert_not_in_migration_rewrite_block(ctx, ql)
+
+            # Start a transaction if we aren't in one already
+            if current_tx.is_implicit() and not in_script:
+                savepoint_name = None
+                tx_cmd = qlast.StartTransaction()
+                tx_query = self._compile_ql_transaction(ctx, tx_cmd)
+                query = dbstate.MigrationControlQuery(
+                    sql=tx_query.sql,
+                    action=dbstate.MigrationAction.START,
+                    tx_action=tx_query.action,
+                    cacheable=False,
+                    modaliases=None,
+                    single_unit=tx_query.single_unit,
+                )
+            else:
+                savepoint_name = current_tx.start_migration()
+                query = dbstate.MigrationControlQuery(
+                    sql=(b'SELECT LIMIT 0',),
+                    action=dbstate.MigrationAction.START,
+                    tx_action=None,
+                    cacheable=False,
+                    modaliases=None,
+                )
+
+            # Start from an empty schema except for `module default`
```

Comment:
Hm, why special treatment for `module default`?

Reply:
A newly created database starts off containing a `module default`, but an empty `FlatSchema()` doesn't.